Code/GraphMol/MolDraw2D/DrawMol.cpp

```diff
@@ -2053,16 +2053,37 @@ void DrawMol::makeBondHighlightLines(double lineWidth, double scale) {
           std::vector<Point2D> end2points;
           makeHighlightEnd(nbr, atom, lineWidth, nbrHighNbrs, end2points);
           std::vector<Point2D> points(end1points);
-          if (end1points.size() > 1 && end2points.size() > 1) {
-            // If the two ends of the polygon point in opposite
-            // directions swap the 2nd end so the path runs correctly.
-            auto v1 = end1points.front().directionVector(end1points.back());
-            auto v2 = end2points.front().directionVector(end2points.back());
-            if (v1.dotProduct(v2) > 0.0) {
-              std::reverse(end2points.begin(), end2points.end());
-            }
-          }
           points.insert(points.end(), end2points.begin(), end2points.end());
+          // the end points are sometimes swapped round, such that a
+          // butterfly-type shape is produced rather than a rectangle
+          // (see Github5592).  Make a convex hull, using a simplified
+          // form of Graham's scan algorithm - all the points
+          // are in the convex hull so it's easier.
+          // sort so the lowest y point is first, with lowest x as tie-breaker
+          std::sort(points.begin(), points.end(),
+                    [](Point2D &p1, Point2D &p2) -> bool {
+                      if (p1.y < p2.y) {
+                        return true;
+                      } else if (p1.y == p2.y) {
+                        return p1.x < p2.x;
+                      }
+                      return false;
+                    });
+          // now sort points 1 -> end so they are all anti-clockwise
+          // around points[0] by checking cross products.
+          std::sort(points.begin() + 1, points.end(),
+                    [&](Point2D &p1, Point2D &p2) -> bool {
+                      auto &p0 = points.front();
+                      auto val = (p1.y - p0.y) * (p2.x - p1.x) -
+                                 (p1.x - p0.x) * (p2.y - p1.y);
+                      if (val == 0.0) {
+                        return (p0 - p2).lengthSq() < (p0 - p1).lengthSq();
+                      } else if (val < 0.0) {
+                        return true;
+                      } else {
+                        return false;
+                      }
```

Comment:
A question: you're doing the comparisons here assuming exact math - i.e. `val == 0.0` or `p1.y < p2.y` - but you are working with doubles; would it be better to use approximate comparisons? 
I'm not sure if it's necessary, but figured it's worth thinking about.

Reply:
That is an interesting question.  Graham's scan normally has a second step that goes round turfing out all the inner points so the convex hull doesn't include them, and I think that that takes care of those sorts of issues.  In this case, we know we are going to have at most a hexagon with defined angles (long with pointy ends) so I don't think the problem will arise.  I can put the 2nd stage in if you like, or document its absence.